File: src/org/andstatus/app/account/AndroidAccountData.java
/**
 * Copyright (C) 2014 yvolk (Yuri Volkov), http://yurivolkov.com
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.andstatus.app.account;

import android.accounts.Account;
import android.accounts.AccountManager;
import android.text.TextUtils;

import org.andstatus.app.util.MyLog;
import org.andstatus.app.util.SharedPreferencesUtil;

class AndroidAccountData implements AccountDataReader {
    private AccountManager am;
    private Account androidAccount;

    AndroidAccountData(AccountManager am, Account androidAccount) {
        this.am = am;
        this.androidAccount = androidAccount;
    }
    
    @Override
    public int getDataInt(String key, int defValue) {
        int value = defValue;
        try {
            String str = getDataString(key, "null");
            if (str.compareTo("null") != 0) {
                value = Integer.parseInt(str);
            }
        } catch (Exception e) {
            MyLog.v(this, e);
        }
        return value;
    }

    long getDataLong(String key, long defValue) {
        long value = defValue;
        try {
            String str = getDataString(key, "null");
            if (str.compareTo("null") != 0) {
                value = Long.parseLong(str);
            }
        } catch (Exception e) {
            MyLog.v(this, e);
        }
        return value;
    }

    boolean getDataBoolean(String key, boolean defValue) {
        boolean value = defValue;
        try {
            String str = getDataString(key, "null");
            if (str.compareTo("null") != 0) {
                value = SharedPreferencesUtil.isTrue(str);
            }
        } catch (Exception e) {
            MyLog.v(this, e);
        }
        return value;
    }
    
    @Override
    public boolean dataContains(String key) {
        boolean contains = false;
        try {
            String str = getDataString(key, "null");
            if (str.compareTo("null") != 0) {
                contains = true;
            }
        } catch (Exception e) {
            MyLog.v(this, e);
        }
        return contains;
    }

    /**
     * User Data associated with the account
     */
    @Override
    public String getDataString(String key, String defValue) {
        String value = defValue;
        String str = am.getUserData(androidAccount, key);
        if (!TextUtils.isEmpty(str)) {
            value = str;
        }
        return value;
    }

    void moveStringKeyTo(String key, AccountDataWriter accountData) {
        accountData.setDataString(key, getDataString(key, null));
        am.setUserData(androidAccount, key, null);
    }

    void moveLongKeyTo(String key, AccountDataWriter accountData) {
        accountData.setDataLong(key, getDataLong(key, 0L));
        am.setUserData(androidAccount, key, null);
    }
}

File: app/src/main/java/org/andstatus/app/data/converter/AndroidAccountData.kt
/*
 * Copyright (C) 2014 yvolk (Yuri Volkov), http://yurivolkov.com
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.andstatus.app.data.converter

import android.accounts.Account
import android.accounts.AccountManager
import org.andstatus.app.account.AccountDataReader
import org.andstatus.app.account.AccountDataWriter
import org.andstatus.app.util.MyLog
import org.andstatus.app.util.SharedPreferencesUtil

internal class AndroidAccountData(private val am: AccountManager?, private val androidAccount: Account?) : AccountDataReader {
    override fun getDataInt(key: String?, defValue: Int): Int {
        var value = defValue
        try {
            val str = getDataString(key, "null")
            if (str.compareTo("null") != 0) {
                value = str.toInt()
            }
        } catch (e: Exception) {
            MyLog.v(this, e)
        }
        return value
    }

    fun getDataLong(key: String?, defValue: Long): Long {
        var value = defValue
        try {
            val str = getDataString(key, "null")
            if (str.compareTo("null") != 0) {
                value = str.toLong()
            }
        } catch (e: Exception) {
            MyLog.v(this, e)
        }
        return value
    }

    fun getDataBoolean(key: String?, defValue: Boolean): Boolean {
        var value = defValue
        try {
            val str = getDataString(key, "null")
            if (str.compareTo("null") != 0) {
                value = SharedPreferencesUtil.isTrue(str)
            }
        } catch (e: Exception) {
            MyLog.v(this, e)
        }
        return value
    }

    override fun dataContains(key: String?): Boolean {
        var contains = false
        try {
            val str = getDataString(key, "null")
            if (str.compareTo("null") != 0) {
                contains = true
            }
        } catch (e: Exception) {
            MyLog.v(this, e)
        }
        return contains
    }

    /**
     * Actor's and User Data associated with the account
     */
    override fun getDataString(key: String?, defValue: String?): String? {
        var value = defValue
        val str = am.getUserData(androidAccount, key)
        if (!str.isNullOrEmpty()) {
            value = str
        }
        return value
    }

    fun moveStringKeyTo(key: String?, accountData: AccountDataWriter?) {
        accountData.setDataString(key, getDataString(key, null))
        am.setUserData(androidAccount, key, null)
    }

    fun moveLongKeyTo(key: String?, accountData: AccountDataWriter?) {
        accountData.setDataLong(key, getDataLong(key, 0L))
        am.setUserData(androidAccount, key, null)
    }
}